8e464c4 breaks clang 12 -O2 (make test failed)

### Summary
seems that 8e464c407a1ce8b88412c6a8cc8aafa8d2cf1b0f (bisected) confused clang 12 when you compile with -O2....

### System information

Mbed TLS version (number or commit id): 3.0.0 or 2.27 branch
Operating system and version: archlinux x86_64 with clang 12, happens with llvm-mingw (windows cross-compiler) too
Configuration (if not default, please attach `mbedtls_config.h`): default
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): 
Additional environment information: 

### Expected behavior

make test succeeded

### Actual behavior
make test failed

### Steps to reproduce
CC=clang cmake .. -DCMAKE_BUILD_TYPE=Release
make -j14
make -j14 test

### Additional information

I can reproduce the failure with Ubuntu clang version 12.0.1-++20210630032618+fed41342a82f-1~exp1~20210630133332.127 from the LLVM apt repository as well as with Ubuntu clang version 12.0.0-1ubuntu1 from Ubuntu 21.04, both on x86_64 and compiling with `-O2`. I can't reproduce it with Clang 11, with `-O1` or for x86_32.

Thanks for doing the bisection. I'm staring at this patch and I really can't see how it might cause undefined behavior. So this is looking like a compiler bug.

Based on the logs from `test_suite_mpi`, it seems that `mpi_mul_hlp` works correctly when called from `mbedtls_mpi_mul_mpi` or `mbedtls_mpi_mul_int`, but not when called by `mpi_montmul`, so exponentiation calculates incorrect result (and then of course so do things that depend on it).
```
Base test mbedtls_mpi_exp_mod #1 .................................. FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Base test mbedtls_mpi_exp_mod #2 (Even N) ......................... PASS
Base test mbedtls_mpi_exp_mod #2 (N = 0 (null)) ................... PASS
Base test mbedtls_mpi_exp_mod #3 (Negative N) ..................... PASS
Base test mbedtls_mpi_exp_mod #4 (Negative base) .................. FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Base test mbedtls_mpi_exp_mod #5 (Negative exponent) .............. PASS
Base test mbedtls_mpi_exp_mod #6 (Negative base + exponent) ....... PASS
Test mbedtls_mpi_exp_mod: 0 (null) ^ 0 (null) mod 9 ............... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 0 (null) ^ 0 (1 limb) mod 9 ............. FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 0 (null) ^ 1 mod 9 ...................... PASS
Test mbedtls_mpi_exp_mod: 0 (null) ^ 2 mod 9 ...................... PASS
Test mbedtls_mpi_exp_mod: 0 (1 limb) ^ 0 (null) mod 9 ............. FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 0 (1 limb) ^ 0 (1 limb) mod 9 ........... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 0 (1 limb) ^ 1 mod 9 .................... PASS
Test mbedtls_mpi_exp_mod: 0 (1 limb) ^ 2 mod 9 .................... PASS
Test mbedtls_mpi_exp_mod: 1 ^ 0 (null) mod 9 ...................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 4 ^ 0 (null) mod 9 ...................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 10 ^ 0 (null) mod 9 ..................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 1 ^ 0 (1 limb) mod 9 .................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 4 ^ 0 (1 limb) mod 9 .................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: 10 ^ 0 (1 limb) mod 9 ................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod: MAX_SIZE exponent ....................... PASS
Test mbedtls_mpi_exp_mod: MAX_SIZE + 1 exponent ................... PASS
Test mbedtls_mpi_exp_mod: MAX_SIZE modulus ........................ PASS
Test mbedtls_mpi_exp_mod: MAX_SIZE + 1 modulus .................... PASS
Test mbedtls_mpi_exp_mod: MAX_SIZE exponent and modulus ........... PASS
Test mbedtls_mpi_exp_mod: MAX_SIZE + 1 exponent and modulus ....... PASS
Test mbedtls_mpi_exp_mod #1 ....................................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod (Negative base) [#1] ..................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_exp_mod (Negative base) [#2] ..................... FAILED
  mbedtls_mpi_cmp_mpi( &Z, &X ) == 0
  at line 1009, ../source/tests/suites/test_suite_mpi.function
…
Base test mbedtls_mpi_is_prime #1 ................................. PASS
Base test mbedtls_mpi_is_prime #2 ................................. PASS
Base test mbedtls_mpi_is_prime #3 ................................. PASS
Base test mbedtls_mpi_is_prime #4 ................................. PASS
Base test mbedtls_mpi_is_prime #5 ................................. PASS
Base test mbedtls_mpi_is_prime #6 ................................. PASS
Base test mbedtls_mpi_is_prime #7 ................................. PASS
Base test mbedtls_mpi_is_prime #8 ................................. PASS
Test mbedtls_mpi_is_prime #1a ..................................... PASS
Test mbedtls_mpi_is_prime #1b ..................................... PASS
Test mbedtls_mpi_is_prime #2a ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #2b ..................................... PASS
Test mbedtls_mpi_is_prime #3 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #4 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #5 [#1] ................................. FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #5 [#2] ................................. FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #6 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #7 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #8 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #9 ...................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #10 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #11 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #12 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #13 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #14 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #15 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #16 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #17 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #18 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #19 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime #20 ..................................... FAILED
  res == div_result
  at line 1104, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime_det (4 non-witnesses) ................... FAILED
  res == 0
  at line 1128, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_is_prime_det (39 non-witnesses) .................. FAILED
  res == 0
  at line 1128, ../source/tests/suites/test_suite_mpi.function
Test mbedtls_mpi_gen_prime (Too small) ............................ PASS
Test mbedtls_mpi_gen_prime (OK, minimum size) ..................... PASS
Test mbedtls_mpi_gen_prime (corner case limb size -1 bits) ........ ^Cmake: *** [Makefile:17090: tests/test_suite_mpi.run] Interrupt
```

--- a/ChangeLog.d/muladdc-amd64-memory.txt
+++ b/ChangeLog.d/muladdc-amd64-memory.txt
@@ -1,0 +1,3 @@
+Bugfix
+   * Fix missing constraints on x86_64 assembly code for bignum multiplication
+     that broke some bignum operations with (at least) Clang 12. Fixes #4786.

--- a/ChangeLog.d/muladdc-amd64-memory.txt
+++ b/ChangeLog.d/muladdc-amd64-memory.txt
@@ -1,3 +1,4 @@
 Bugfix
    * Fix missing constraints on x86_64 assembly code for bignum multiplication
-     that broke some bignum operations with (at least) Clang 12. Fixes #4786.
+     that broke some bignum operations with (at least) Clang 12.
+     Fixes #4116, #4786, #4917.

--- a/library/bn_mul.h
+++ b/library/bn_mul.h
@@ -225,9 +225,9 @@
         "addq   $8, %%rdi\n"
 
 #define MULADDC_STOP                        \
-        : "+c" (c), "+D" (d), "+S" (s)      \
-        : "b" (b)                           \
-        : "rax", "rdx", "r8"                \
+        : "+c" (c), "+D" (d), "+S" (s), "+m" (*(uint64_t (*)[16]) d) \
+        : "b" (b), "m" (*(const uint64_t (*)[16]) s)                 \
+        : "rax", "rdx", "r8"                                         \
     );
 
 #endif /* AMD64 */

